### PR TITLE
[Blocking][jvm-packages] For training data with group, empty RDD partition threw exception (#3749)

### PR DESCRIPTION
**Pull request by weitian**
private var isNewGroup = true

  override def hasNext: Boolean = {
    return base.hasNext || isNewGroup
 }

The initial value for isNewGroup should be false otherwise the hasNext is wrong with empty partition.

### CODE REVIEW

**Comment by hcho3** [comment]
Can you put your handle to the list of contributors in CONTRIBUTORS.md?

**Comment by CodingCat** [comment]
can we have a test covering that?

BTW, since we have some kind of special logic in ranker training now, I think it makes sense to have a separate suite for all ranking tests

**Comment by weitian** [comment]
> can we have a test covering that?
> 
> BTW, since we have some kind of special logic in ranker training now, I think it makes sense to have a separate suite for all ranking tests

Yes, we should add a XGBoostRankerSuite with more test cases.

For ranker training, "rank:pairwise" is a major reason we chose XGBoost over other machine learning platform.
I suggest we could add a XGBoostRanker to wrap all ranker related logic, too.

This could be a big topic all together, I suggest we create another issue and work on it? what do you think?

**Comment by hcho3** [comment]
I'm curious, is rank:ndcg and rank:map supported in XGBoost4J-Spark? If so, how do they measure in practice compared to rank:pairwise?

**Comment by CodingCat** [comment]
I think XGBoostRanker might be too large for a PR right before the release, but it doesn't prevent it to be a potentially good idea, 

But before you code I would like to discuss over a brief plan, so that we can prevent it from being an over-design...

**Comment by CodingCat** [comment]
for this PR, we can focus on fix the issue and have a test guarding that

**Comment by CodingCat** [comment]
@hcho3 yes, I think we have supported all objective functions https://github.com/dmlc/xgboost/blob/1878e0b6d152ca7bb643e94b7d20f2331cc2ffc3/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/params/LearningTaskParams.scala#L94-L102

**Comment by hcho3** [comment]
@CodingCat @weitian I am thinking about migrating some tests from Travis to Jenkins, since they tend to fail due to lack of memory or connection problems.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala [hunk 1]
```diff
@@ -267,4 +268,14 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
 
     assert(booster != null)
   }
+
+  test("repartitionForTrainingGroup with empty partition") {
+    val trainingRDD = sc.parallelize(Ranking.train, 5).mapPartitions(it => {
+      // make one partition empty for testing
+      it.filter(_ => TaskContext.getPartitionId() != 3)
+    })
+    val traingGroupsRDD = XGBoost.repartitionForTrainingGroup(trainingRDD, 4)
```
Comment (CodingCat) [inline jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: can you call through trainDistributed API, which is the unified interface for XGBClassifier/Regressor 

and you may want to change repartitionForTrainingGroup to private

Comment (weitian) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: changed the test to call trainDistributed API.

repartitionForTrainingGroup is package level private already just like the repartitionForTraining method.
API user have no access to it, but it is needed to write unit test.

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: as the current unit test is kind of coarse-grained, i.e. we didn't test the most nested call of the function, but test everything along the calling path by calling trainDistributed()

I just noticed that the previous ranking trainer test are also calling repartitionForTrainingGroup()

so you should either change everything to call trainDistributed() and make repartitionForTrainingGroup as private, 

or you should have two unit test cases for everything you want to test, by calling trainDistributed and repartitionForTrainingGroup respecitively

Comment (weitian) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: > as the current unit test is kind of coarse-grained, i.e. we didn't test the most nested call of the function, but test everything along the calling path by calling trainDistributed()
> 
> I just noticed that the previous ranking trainer test are also calling repartitionForTrainingGroup()
> 
> so you should either change everything to call trainDistributed() and make repartitionForTrainingGroup as private,
> 
> or you should have two unit test cases for everything you want to test, by calling trainDistributed and repartitionForTrainingGroup respectively

1. There is another unit test calling trainDistributed(): test("distributed training with group data")
2. repartitionForTrainingGroup was defined as 'private[spark]' which follows the convention of the existing method 'private[spark] def repartitionForTraining()'
3. If you take a close look of 'test("repartitionForTrainingGroup with group data")', there is no way to verify whether the grouped data is correct or not by calling  trainDistributed() instead of repartitionForTrainingGroup()

Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: the existence of this bug shows that test("distributed training with group data") doesn't cover the test with data containing empty partition, 

test("repartitionForTrainingGroup with group data")' should be changed to something like test("distributed training with group data containing empty partition")' 

because we have to ensure that  the code path (and the future change on them) in trainDistributed() before  `repartitionForTrainingGroup` has no impact on the things you are testing here... trainDistributed() is actually the entry for everything outside of this file 


if you intentionally want to test the data integrity produced by  `repartitionForTrainingGroup` keep it as private[spark] is fine and you may want to add back your original test in addition to the current one

Comment (weitian) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: add the original test back.

1. test("repartitionForTrainingGroup with group data"): test data integrity after re-grouping.
2. test("repartitionForTrainingGroup with group data which has empty partition"): test corner case with empty partition.
3. test("distributed training with group data which has empty partition"): test from trainDistributed() which is the API to the outside.

**Comment by CodingCat** [comment]
LGTM, merged

**Comment by CodingCat** [comment]
Thanks
